tests/gas_subsidies/deducts_funds_test.go

```diff
@@ -0,0 +1,372 @@
+// Copyright 2025 Sonic Operations Ltd
+// This file is part of the Sonic Client
+//
+// Sonic is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Sonic is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Sonic. If not, see <http://www.gnu.org/licenses/>.
+
+package gas_subsidies
+
+import (
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/0xsoniclabs/sonic/gossip/blockproc/subsidies"
+	"github.com/0xsoniclabs/sonic/gossip/blockproc/subsidies/registry"
+	"github.com/0xsoniclabs/sonic/opera"
+	"github.com/0xsoniclabs/sonic/tests"
+	"github.com/0xsoniclabs/sonic/tests/contracts/counter"
+	"github.com/0xsoniclabs/sonic/tests/contracts/revert"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGasSubsidies_SubsidizedTransaction_DeductsSubsidyFunds(t *testing.T) {
+	upgrades := []struct {
+		name    string
+		upgrade opera.Upgrades
+	}{
+		{name: "sonic", upgrade: opera.GetSonicUpgrades()},
+		{name: "allegro", upgrade: opera.GetAllegroUpgrades()},
+		// TODO: add brio once it supports internal transactions
+	}
+	singleProposerOption := map[string]bool{
+		"singleProposer": true,
+		"distributed":    false,
+	}
+
+	for _, test := range upgrades {
+		for mode, enabled := range singleProposerOption {
+			t.Run(fmt.Sprintf("%s/%v", test.name, mode), func(t *testing.T) {
+
+				test.upgrade.GasSubsidies = true
+				test.upgrade.SingleProposerBlockFormation = enabled
+				net := tests.StartIntegrationTestNet(t, tests.IntegrationTestNetOptions{
+					Upgrades: &test.upgrade,
+				})
+
+				testGasSubsidies_SubsidizedTransaction_DeductsSubsidyFunds(t, net)
+			})
+		}
+	}
+}
+
+func testGasSubsidies_SubsidizedTransaction_DeductsSubsidyFunds(t *testing.T, net *tests.IntegrationTestNet) {
+
+	client, err := net.GetClient()
+	require.NoError(t, err)
+	defer client.Close()
+
+	revertContract, receipt, err := tests.DeployContract(net, revert.DeployRevert)
+	require.NoError(t, err)
+	require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+
+	counterContract, receipt, err := tests.DeployContract(net, counter.DeployCounter)
+	require.NoError(t, err)
+	require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+
+	cases := map[string]struct {
+		runTransactions func(t *testing.T, net *tests.IntegrationTestNet, sender *tests.Account)
+	}{
+		"sponsored transaction calls contract": {
```

Comment:
A case probably to be covered in another test:
- a transaction creating a contract (To=nil, gasPrice=0) -- should be rejected